### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by humanpose1**
add a `BaseTracker `and `SegmentationTracker` add test for confusion matrix and segmentation tracker. Test for `SegmentationTracker` still not pass.

### CODE REVIEW

**Comment by humanpose1** [comment]
I think the trackers could also have a config with hydra.

**Comment by clee-ai** [comment]
I also wonder if we could utilize a more lightning-friendly library to do some of the metrics for us, maybe https://torchmetrics.readthedocs.io/en/latest/ ? (we can do that later)

**Comment by humanpose1** [comment]
Yes You're right. The goal is to use torchMetrics when we can. And here I use torchmetrics to compute the confusion matrix. But for S3DIS, Scannet, or shapenet I prefer staying close the the implementation of Nicolas Chaulet, because these metrics are specific for their respective benchmark. It would be cumbersome to have different results with the new metrics.

The `BaseTracker` tracks the loss also in our implementation.

**Comment by humanpose1** [comment]
@CCInc Can you have a look at `BaseTracker` ? Do you agree to process loss in the tracker ? Or should we do something like lightning transformer ?

Also, the input and the output of the tracker are dict of tensor.
Now the tests pass (you need to install pytest).

**Comment by humanpose1** [comment]
use this command
```
python -m pytest test

```

**Comment by clee-ai** [comment]
I wonder if we should do something more similar to lightning transformers, where metrics are associated with models rather than calculated separately. https://github.com/PyTorchLightning/lightning-transformers/blob/master/lightning_transformers/task/nlp/multiple_choice/model.py#L55

Then, we can provide our own metrics which subclass `torchmetrics.Metric`, and define the metrics common to all segmentation tasks in `SegmentationBaseModel`. This way, the user has more control over the metrics for each model, so to change metrics for a model, they just have to override `configure_metrics` rather than change the tightly coupled tracker file.

**Comment by humanpose1** [comment]
are you thinking of something like:
```python

class OverallAccuracy(Metric):
...

class AverageIOU(Metric):
...

```
and then in `configure_metrics` we would have something like:
```python
def configure_metrics(self):
acc = OverallAccuracy(num_classes=num_classes)
iou = AverageIOU(num_classes=num_classes)
```
Is it what you are thinking about ?

In our case, we use `torchmetrics.ConfusionMatrix`. and with the confusion matrix, we can compute the overall accuracy, the mean class accuracy,  the mean iou, the iou per class. In the case described above, we'll compute the confusion matrix several times. 
Here it's okay because computing a confusion matrix is fast. But it is not always the case.

**Comment by clee-ai** [comment]
Yes, something like that.

Instead of using the confusion matrix, can we instead calculate the metric directly, such as in torchmetrics? https://github.com/PyTorchLightning/metrics/blob/master/torchmetrics/functional/classification/dice.py and that way we avoid some of the computational cost hopefully.

**Comment by humanpose1** [comment]
computing a confusion matrix is not costly (it is a matrix of `num_classes` x `num_classes`). For nlp or for classification/segmentation, it seems to be a good solution.
 But for other tasks such as registration, we need to compute intermediate results and it can be costly. 
For registration, 
1) we compute descriptors for two scenes using the neural network
2) we compute matches (costly) between the scenes.
3a) with matches, we compute metrics(feature match recall, hit ratio etc)
3b) with matches, we compute a transformation using a robust estimator (RANSAC or Fast global registration) (costly)
4) with the transformation, we compute transformation metrics (rotation error, translation error, chamfer distance...).

In this scenario, we cannot implement a class for each metric and apply it independently in a `compute_metrics` method, it would be inefficient.
That is why, I prefer keeping a tracker.

**Comment by clee-ai** [comment]
Sounds good, let's keep it the same then for now!

File: test/test_model.py [hunk 1]
```diff
@@ -9,27 +9,25 @@
...
 
 
-class TestAPIModel(unittest.TestCase):
-    def test_forward(self):
-        option_dataset = OmegaConf.create({"feature_dimension": 1, "num_classes": 10})
+@pytest.mark.skip("For now we skip the tests...")
+def test_forward(self):
+    option_dataset = OmegaConf.create({"feature_dimension": 1, "num_classes": 10})
+    option_criterion = OmegaConf.create({"_target_": "torch.nn.NLLLoss"})
+    instantiator = HydraInstantiator()
```
Comment (clee-ai) [inline test/test_model.py:22]: Once we get the structuredconfig working, we should be able to instantiate the config objects directly from the dataclasses with defaults, which will make testing a lot easier.

File: torch_points3d/models/segmentation/base_model.py [hunk 2]
```diff
@@ -28,17 +28,25 @@ def set_input(self, data: Data) -> None:
...
-    def get_losses(self) -> Union[torch.Tensor, None]:
-        # only compute loss if loss is defined and the dset has labels
-        if self.labels is None or self.criterion is None:
-            return
-
-        self.loss = self.criterion(self.output, self.labels)
-        return self.loss
+        self._output = F.log_softmax(logits, dim=-1)
+        self.compute_losses()
+        if "loss" in self._losses.keys():
```
Comment (clee-ai) [inline torch_points3d/models/segmentation/base_model.py]: I think this is too complicated. We should just `return self.compute_losses()` and have the global loss calculation isolated to compute_losses. I want to keep the forward as simple as possible.

File: torch_points3d/models/segmentation/base_model.py [hunk 3]
```diff
@@ -28,17 +28,25 @@ def set_input(self, data: Data) -> None:
...
+            return self._losses["loss"]
+        else:
+            return None
+
+    def compute_losses(self):
```
Comment (clee-ai) [inline torch_points3d/models/segmentation/base_model.py:38]: have this function return the global loss

**Comment by humanpose1** [comment]
I've solved the problem of metrics it does not display tensor anymore

File: torch_points3d/metrics/segmentation/segmentation_tracker.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
+import torch
+from typing import Dict, Optional, Tuple, Any, Union
+from torchmetrics import ConfusionMatrix
+from torchmetrics import Metric
+
+from torch_points3d.metrics.base_tracker import BaseTracker
+
+
+def compute_average_intersection_union(confusion_matrix: torch.Tensor, missing_as_one: bool = False) -> torch.Tensor:
```
Comment (clee-ai) [inline torch_points3d/metrics/segmentation/segmentation_tracker.py]: can we move these definitions to another file? maybe torch_points3d/core/metrics.py?

File: torch_points3d/metrics/segmentation/segmentation_tracker.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        self._num_classes = num_classes
+        self.confusion_matrix_metric = ConfusionMatrix(num_classes=self._num_classes)
+        self.eps = eps
+
+    def compute_metrics_from_cm(self, matrix: torch.Tensor) -> Dict[str, Any]:
+        acc = compute_overall_accuracy(matrix)
+        macc = compute_mean_class_accuracy(matrix)
+        miou = compute_average_intersection_union(matrix)
+        iou_per_class, _ = compute_intersection_union_per_class(matrix, eps=self.eps)
+        iou_per_class_dict = {f"{self.stage}_iou_class_{i}": (100 * v) for i, v in enumerate(iou_per_class)}
```
Comment (clee-ai) [inline torch_points3d/metrics/segmentation/segmentation_tracker.py:33]: I think here is where we need to add `.item()`

`{f"{self.stage}_iou_class_{i}": (100 * v.item()) for i, v in enumerate(iou_per_class)}`

File: torch_points3d/metrics/base_tracker.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class BaseTracker(nn.Module):
+    """
+    pytorch Module to manage the losses and the metrics
+    """
+
+    def __init__(self, stage: str = "train"):
+        super().__init__()
+        self.stage: str = stage
```
Comment (clee-ai) [inline torch_points3d/metrics/base_tracker.py:14]: I think if we set the stage here, we should remove all the arguments asking for the stage in the functions.

File: torch_points3d/metrics/base_tracker.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        return final_metrics
+
+    def get_final_loss_metrics(self):
+        metrics = dict()
+        for key, m in self.loss_metrics.items():
+            metrics[key] = m.compute()
+        self.loss_metrics = nn.ModuleDict()
+        return metrics
+
+    def reset(self, stage: str = "train"):
```
Comment (clee-ai) [inline torch_points3d/metrics/base_tracker.py]: remove stage param?

File: torch_points3d/models/base_model.py [hunk 8]
```diff
@@ -12,12 +12,28 @@ def __init__(self, instantiator: Instantiator):
...
+    def compute_loss(self):
         raise (NotImplementedError("get_losses needs to be defined!"))
+
+    def get_losses(self, stage: Optional[str] = None) -> Optional[Dict["str", torch.Tensor]]:
+        if stage is None:
+            return self._losses
+        else:
+            losses = {}
+            for name, l in self._losses.items():
+                losses[f"{stage}_{name}"] = l.item()
```
Comment (clee-ai) [inline torch_points3d/models/base_model.py]: I feel like this is outside of the scope of the model. We should simply return the model losses without worrying about what stage the model is, and handle the stage upstream.

**Review by clee-ai — COMMENTED** [review]
It's the right idea! I'm trying to think if there's a better way to integrate between LitModel/Tracker/Model but I can't come up with one. It's difficult to find a good mix of abstraction/usability with libraries like this, so I think what you did will work well.

I don't like having to store all the losses for a model in a dictionary, but I don't know how else we'd be able to return them to the tracker when dealing with multiple losses. I'm also unsure about returning a dictionary for `get_outputs` but I understand why it is that way, and I again have no better solution 😄 

I think we should work hard to keep the `BaseModel` as simple as possible, i.e. not having to worry about trackers and stages such. We should try to abstract away as much as possible to the Tracker.

File: torch_points3d/metrics/segmentation/segmentation_tracker.py (hunk 5 above)
Comment (clee-ai) [reply torch_points3d/metrics/segmentation/segmentation_tracker.py:33]: I see you fixed it, nevermind!

File: torch_points3d/models/base_model.py (hunk 8 above)
Comment (humanpose1) [reply torch_points3d/models/base_model.py]: You're right. It's fixed

File: torch_points3d/metrics/base_tracker.py (hunk 7 above)
Comment (humanpose1) [reply torch_points3d/metrics/base_tracker.py]: In that case, we can remove reset method

File: torch_points3d/metrics/segmentation/segmentation_tracker.py (hunk 4 above)
Comment (humanpose1) [reply torch_points3d/metrics/segmentation/segmentation_tracker.py]: Done

File: torch_points3d/models/segmentation/base_model.py (hunk 3 above)
Comment (humanpose1) [reply torch_points3d/models/segmentation/base_model.py:38]: Done

File: torch_points3d/models/segmentation/base_model.py (hunk 2 above)
Comment (humanpose1) [reply torch_points3d/models/segmentation/base_model.py]: Done

File: torch_points3d/metrics/base_tracker.py (hunk 7 above)
Comment (humanpose1) [reply torch_points3d/metrics/base_tracker.py]: Done

**Comment by humanpose1** [comment]
If we want to adapt to many conditions, we'll need to return something like a dictionary. But maybe we can have a specific class for the losses and the outputs of the network if we want to check the type for example.

**Comment by clee-ai** [comment]
Great! It looks good now, we can merge whenever you want. 

And I agree, this is a case of being difficult to write code that is both clean, modular, and adaptable. I think it makes me uncomfortable because, if the user changes `get_outputs`, it'll be difficult to remember to also change `track`... maybe there's something we can do in the future with using dataclasses to pass outputs, since `get_outputs` in the model and `track` in the tracker will always be related (segmentation tracker goes with segmentation model, so we have one dataclass for outputs; registration tracker goes with registration model, so we have a different dataclass for those outputs, etc.) but we can look at that later.
